lib/App/DuckPAN/Cmd/New.pm

```diff
@@ -44,40 +62,82 @@ sub run {
-		test       => ["$filepath.t"],
-		code       => ["$filepath.pm"],
-		handlebars => [$lc_filepath, "$lc_name.handlebars"],
-		js         => [$lc_filepath, "$lc_name.js"]);
 	my %vars = (
 		ia_name           => $name,
 		ia_package_name   => $package_name,
 		ia_name_separated => $separated_name,
 		lia_name          => $lc_name,
-		ia_path           => $filepath
+		ia_path           => $filepath,
+		# TODO Remove this before merging to master
+		lia_id            => $lc_name,
 	);
-	foreach my $template_type (sort keys %template_info) {
-		my ($source, $dest) = ($template_info{$template_type}{in}, $template_info{$template_type}{out});
-		$self->app->emit_and_exit(-1, 'Template does not exist: ' . $source) unless ($source->exists);
-		# Update dest based on type:
-		$dest = $dest->child(@{$files{$template_type}});
-		$self->app->emit_and_exit(-1, 'File already exists: "' . $dest->basename . '" in ' . $dest->parent) if ($dest->exists);
-		my $content = $tx->render("$source", \%vars);
-		$dest->touchpath->append_utf8($content);    #create file path and append to file
-		$self->app->emit_info("Created file: $dest");
+
+	# Ask which optional groups to create
+	my @optionals;
+
+	for my $optional (@{$template->optional_groups}) {
+		my $reply = $self->app->get_reply('Do you want to create files for "' . $optional . '" [y/n]? ');
+
+		push @optionals, $optional if $reply =~ /^y/i;
 	}
+
+	# Create the files
+	my @created_files;
+
+	try {
+		@created_files = $template->generate(\%vars, \@optionals);
+	} catch {
+		$self->app->emit_and_exit(-1,
+			"Error processing template '" . $template->name . "' $_");
+	};
+
+	$self->app->emit_info("Created file: $_") for sort @created_files;
 	$self->app->emit_info("Successfully created " . $type->{name} . ": $package_name");
 }
 
+# Process the template related arguments and return the chosen template
+sub _process_template_args {
+	my $self = shift;
+	my $type = $self->app->get_ia_type();
+	my $template_defs;
+
+	# read the templates.yml file
+	try {
+		$template_defs = App::DuckPAN::TemplateDefinitions->new;
+	} catch {
+		if (/no such file/i) {
+			$self->app->emit_and_exit(-1,
+				"No templates exist for this IA Type: " . $type->{name} . ". " .
+				"Maybe you need to pull the latest version of this repository?");
+		} else {
+			$self->app->emit_and_exit(-1, "Error reading template definitions file: $_");
+		}
+	};
+
+	my $available_templates = join(", ", sort $template_defs->get_template_names());
+
+	# process the --list-templates option: list the templates and exit with success
+	$self->app->emit_and_exit(0, "Available templates are:\n$available_templates")
+		if $self->list_templates;
+
+	my $template = $template_defs->get_template($self->template);
+
+	# oops, the template was not found
+	unless ($template) {
+		$self->app->emit_and_exit(-1,
+			"Template '" . $self->template . "' does not exist for this IA Type: " .
+			$type->{name} . ". Please use one of: " . $available_templates);
```

Comment:
I have updated the _available templates_ messaging. The _description_ key would be nice, but I might leave it for later once we decide on the [new the template file format](https://github.com/duckduckgo/zeroclickinfo-goodies/pull/1778), because it is not possible to merge this PR before the template definition files have been merged. Thanks!